syncer/pipe.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+// Pipe is the littlest process unit in syncer
+type Pipe interface {
+	// Name is the pipe's name
+	Name() string
+
+	// Init initializes the pipe, iFunc also do some initialize job.
+	Init(cfg *config.SubTaskConfig, iFunc func() error) error
+
+	// Update updates the pipe's config
+	Update(*config.SubTaskConfig) error
+
+	// Input receives data
+	Input() chan *PipeData
+
+	// Output exports data already be processed
+	Output() chan *PipeData
+
+	// Run starts the pipe's process
+	Run()
+
+	// Close closes the pipe
+	Close()
+
+	// Wait waits all the data in the pipe is processed
+	Wait()
+
+	// SetErrorChan set the error channel, send error to this channel when meet error
+	SetErrorChan(chan error)
+
+	// SetResolveFunc set the resolveFunc, which used when data is resolved, used in the last pipe, or data is skiped
+	SetResolveFunc(func())
+}
+
+// PipeData is the data processed in pipe
+type PipeData struct {
+	binlogEvent *replication.Event
+
+	tp             opType
+	sourceSchema   string
+	sourceTable    string
+	targetSchema   string
+	targetTable    string
+	sqls           []string
+	args           [][]interface{}
+	keys           [][]string
+	retry          bool
+	lastPos        mysql.Position
+	currentPos     mysql.Position // exactly binlog position of current SQL
+	currentGTIDSet gtid.Set
+	ddls           []string
+	traceID        string
+	traceGID       string
+
+	ddlExecItem *DDLExecItem
+
+	// may have other data needed
+	additionalData interface{}
+}
+
+// Pipeline contains many pipes, and resolve data by every pipe.
+type Pipeline struct {
+	pipes []Pipe
+
+	closeCh  chan interface{}
+	isClosed sync2.AtomicBool
+
+	errCh chan error
+
+	wg     sync.WaitGroup
+	dataWg sync.WaitGroup
+}
+
+// NewPipeline returns a new pipeline
+func NewPipeline() *Pipeline {
+	return &Pipeline{
+		pipes: make([]Pipe, 0, 5),
+	}
+}
+
+// AddPipe adds a pipe to this pipeline
+func (p *Pipeline) AddPipe(pipe Pipe) {
+	if len(p.pipes) == 0 {
```

Comment:
```suggestion
	if p.pipes == nil {
```
If the `Pipeline` is created by `NewPipeline`, we can simply append pipe to p.pipes